insecure/adversary/attacker_test.go

```diff
@@ -0,0 +1,195 @@
+			// messages conccurently to attacker).
+			attackerSendWG := sync.WaitGroup{}
+			attackerSendWG.Add(concurrencyDegree)
+
+			for _, msg := range messages {
+				msg := msg
+
+				go func() {
+					err := client.Send(msg)
+					require.NoError(t, err)
+					attackerSendWG.Done()
+				}()
+			}
+
+			// all messages should be sent to attacker in a timely fashion.
+			unittest.RequireReturnsBefore(t, attackerSendWG.Wait, 1*time.Second, "could not send all messages to attacker on time")
+			// all events should be relayed to the orchestrator by the attacker in a timely fashion.
+			unittest.RequireReturnsBefore(t, orchestratorWG.Wait, 1*time.Second, "orchestrator could not receive messages on time")
+		})
+}
+
+// messageFixture creates and returns a randomly generated gRPC message that is sent between a corruptible conduit and the attacker.
+// It also generates and returns the corresponding application-layer event of that message, which is sent between the attacker and the
+// orchestrator.
+func messageFixture(t *testing.T, codec network.Codec) (*insecure.Message, *insecure.Event) {
+	// fixture for content of message
+	originId := unittest.IdentifierFixture()
+	targetIds := unittest.IdentifierListFixture(10)
+	targets := uint32(3)
+	protocol := insecure.Protocol_MULTICAST
+	channel := network.Channel("test-channel")
+	content := &message.TestMessage{
+		Text: fmt.Sprintf("this is a test message: %d", rand.Int()),
+	}
+
+	// encodes event to create payload
+	payload, err := codec.Encode(content)
+	require.NoError(t, err)
+
+	// creates message that goes over gRPC.
+	m := &insecure.Message{
+		ChannelID: "test-channel",
+		OriginID:  originId[:],
+		Targets:   targets,
+		TargetIDs: flow.IdsToBytes(targetIds),
+		Payload:   payload,
+		Protocol:  protocol,
+	}
+
+	// creates corresponding event of that message that
+	// is sent by attacker to orchestrator.
+	e := &insecure.Event{
+		CorruptedId: originId,
+		Channel:     channel,
+		Content:     content,
+		Protocol:    protocol,
+		TargetNum:   targets,
+		TargetIds:   targetIds,
+	}
+
+	return m, e
+}
+
+// messageFixtures creates and returns randomly generated gRCP messages and their corresponding protocol-level events.
+// The messages are sent between a corruptible conduit and the attacker.
+// The events are the corresponding protocol-level representation of messages.
+func messageFixtures(t *testing.T, codec network.Codec, count int) ([]*insecure.Message, []*insecure.Event) {
+	msgs := make([]*insecure.Message, count)
+	events := make([]*insecure.Event, count)
+
+	for i := 0; i < count; i++ {
+		m, e := messageFixture(t, codec)
+
+		msgs[i] = m
+		events[i] = e
+	}
+
+	return msgs, events
+}
+
+// withAttacker is a test helper that creates and starts an attacker server, prepares a set of message fixtures and runs the given scenario
+// against the attacker and generated messages.
+func withAttacker(
```

Comment:
can you add more high level explanation of how exactly `withAttacker()` works? I find it confusing to follow the logic of this function to understand what exactly it's doing

Reply:
Tried decomposing the test logic into smaller components in https://github.com/onflow/flow-go/pull/2121/commits/7766095508a35da50eb9f382dd8b1cd3f99e4ef6 and added more comments.